compiler/src/dmd/errors.d

```diff
@@ -35,6 +35,133 @@ enum ErrorKind
     message,
 }
 
+// Struct for SARIF Tool Information
+struct ToolInformation {
+    string name;
+    string toolVersion;
+
+    string toJson() nothrow {
+        return `{"name": "` ~ name ~ `", "version": "` ~ toolVersion ~ `"}`;
+    }
+}
+
+// Struct for SARIF Result
+struct Result {
+    string ruleId;  // Rule identifier
+    string message;  // Error message
+    PhysicalLocation location;  // Location where the error occurred
+
+    string toJson() nothrow {
+        return `{"ruleId": "` ~ ruleId ~ `", "message": "` ~ message ~ `", "location": ` ~ location.toJson() ~ `}`;
+    }
+}
+
+// Struct for Physical Location of the Error
+struct PhysicalLocation {
+    string uri;  // File path (URI)
+    int startLine;  // Line number where the error occurs
+    int startColumn;  // Column number where the error occurs
+
+    string toJson() nothrow {
+        return `{"artifactLocation": {"uri": "` ~ uri ~ `"}, "region": {"startLine": ` ~ intToString(startLine) ~ `, "startColumn": ` ~ intToString(startColumn) ~ `}}`;
+    }
+}
+
+// SARIF Report Struct
+struct SarifReport {
+    ToolInformation tool;  // Information about the tool
+    Invocation invocation;  // Information about the execution
+    Result[] results;  // List of results (errors, warnings, etc.)
+
+    string toJson() nothrow {
+        string resultsJson = "[" ~ results[0].toJson();
+        foreach (result; results[1 .. $]) {
+            resultsJson ~= ", " ~ result.toJson();
+        }
+        resultsJson ~= "]";
+
+        return `{"tool": ` ~ tool.toJson() ~ `, "invocation": ` ~ invocation.toJson() ~ `, "results": ` ~ resultsJson ~ `}`;
+    }
+}
+
+// Struct for Invocation Information
+struct Invocation {
+    bool executionSuccessful;
+
+    string toJson() nothrow {
+        return `{"executionSuccessful": ` ~ (executionSuccessful ? "true" : "false") ~ `}`;
+    }
+}
+
+// Function to replace writeln with fprintf for printing to stdout
+void printToStdout(string message) nothrow {
+    fprintf(stdout, "%.*s\n", cast(int)message.length, message.ptr);  // Cast to int
+}
+
+void generateSarifReport(const ref Loc loc, const(char)* format, va_list ap, ErrorKind kind) nothrow
+{
+    string toolVersion = global.versionString();  // Retrieve dynamic version
+    ToolInformation tool = ToolInformation("DMD", toolVersion);
```

Comment:
This variable is unused, and causes the buildkite CI Style check to fail

Reply:
Got it! I’ll remove the unused variables and rebase the branch as suggested. I’ll also update `frontend.h` after finishing the current changes. Thanks for pointing it out!